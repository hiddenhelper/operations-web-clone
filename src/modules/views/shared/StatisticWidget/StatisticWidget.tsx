import React, { memo } from 'react';

import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import Skeleton from '@material-ui/lab/Skeleton';

import { getConditionalDefaultValue } from '../../../../utils/generalUtils';
import { statusWidgetSkeletonStyles, statusWidgetStyles } from '../DashboardWidget/styles';

export interface IStatisticWidget {
  title: string;
  subtitle?: string;
  content: string | number;
  renderContentPill?: React.ReactNode;
  renderAuxContent?: React.ReactNode;
  renderActivity?: React.ReactNode;
  inlineContent: boolean;
  widgetCta?: React.ReactNode;
  loading?: boolean;
}

const StatisticWidget = ({
  title,
  subtitle,
  content,
  renderContentPill,
  renderAuxContent,
  renderActivity,
  inlineContent,
  widgetCta,
  loading,
}: IStatisticWidget) => {
  const skeletonClasses = statusWidgetSkeletonStyles();
  const classes = statusWidgetStyles();

  return (
    <Paper className={`${classes.widget} ${classes.noDetail}`} elevation={0}>
      {loading && inlineContent && (
        <div>
          <div className={skeletonClasses.wrapper}>
            <Skeleton className={skeletonClasses.inlineMain} animation={'wave'} variant="text" width={270} height={58} />
            {renderActivity && <Skeleton className={skeletonClasses.inlineActivity} animation={'wave'} variant="text" width={270} height={25} />}
          </div>
          {renderAuxContent && <Skeleton className={skeletonClasses.inlineAuxContent} animation={'wave'} variant="rect" width={'100%'} height={216} />}
          <Skeleton className={skeletonClasses.inlineCta} animation={'wave'} variant="text" width={120} height={23} />
        </div>
      )}
      {loading && !inlineContent && (
        <div>
          <Typography className={`${classes.widgetTitle} ${classes.widgetTitleBottomMargin}`}>{title}</Typography>
          <div className={skeletonClasses.columnMainContent} data-testid="statistics-widget-skeleton">
            <Skeleton className={skeletonClasses.columnContent} animation={'wave'} variant="text" width={270} height={63} />
            {renderAuxContent && <Skeleton className={skeletonClasses.columnAuxContent} animation={'wave'} variant="rect" width={'100%'} height={168} />}
          </div>
          {!renderAuxContent && <Skeleton className={skeletonClasses.columnActivity} animation={'wave'} variant="text" width={200} height={23} />}
        </div>
      )}
      {!loading && (
        <>
          <div className={classes.widgetContainer}>
            {inlineContent ? (
              <div>
                <div className={classes.widgetInlineStatusWrapper}>
                  <div
                    className={`${classes.widgetContentWrapper} ${getConditionalDefaultValue(!renderAuxContent, classes.widgetContentWrapperBottomMargin, '')}`}
                  >
                    <span className={`${classes.widgetContent} ${classes.widgetInlineMargin}`}>{content}</span>
                    <Typography className={` ${classes.widgetTitle} ${classes.widgetTitleInlinePosition}`}>{title}</Typography>
                    <div className={classes.chipStatusLeftMargin}>{renderContentPill}</div>
                  </div>
                  <div
                    className={`${getConditionalDefaultValue(
                      !renderAuxContent,
                      classes.widgetMarginBottom + ' ' + classes.widgetContentWrapperNoTopMargin,
                      ''
                    )}`}
                  >
                    {renderActivity}
                  </div>
                </div>
                <Typography className={`${classes.widgetSubtitle} ${classes.widgetSubtitleInlinePosition}`}>{subtitle}</Typography>
              </div>
            ) : (
              <>
                <Typography className={`${classes.widgetTitle} ${classes.widgetTitleBottomMargin}`}>{title}</Typography>
                {subtitle && <Typography className={`${classes.widgetSubtitle} ${classes.widgetSubtitlePosition}`}>{subtitle}</Typography>}
                <div
                  className={`${classes.widgetContentWrapper} ${getConditionalDefaultValue(
                    !subtitle && !renderAuxContent,
                    classes.widgetContentWrapperBottomMargin,
                    classes.widgetTitleMarginBottom
                  )}`}
                >
                  <span className={`${classes.widgetContentResponsive} ${classes.widgetContent}`}>{content}</span>
                  <div className={classes.chipStatusLeftMargin}>{renderContentPill}</div>
                </div>
              </>
            )}

            <div className={`${getConditionalDefaultValue(renderAuxContent, classes.widgetAuxContentPadding, '')}`}>{renderAuxContent}</div>
          </div>

          {!inlineContent && (
            <div className={`${getConditionalDefaultValue(!renderAuxContent, classes.widgetContentWrapperNoTopMargin, '')}`}>{renderActivity}</div>
          )}

          {widgetCta}
        </>
      )}
    </Paper>
  );
};

export default memo(StatisticWidget);
