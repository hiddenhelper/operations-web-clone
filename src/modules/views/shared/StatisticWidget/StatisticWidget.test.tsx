import React from 'react';
import { render } from '@testing-library/react';

import StatisticWidget, { IStatisticWidget } from './StatisticWidget';

describe('StatisticWidget', () => {
  let wrapper;
  let props: IStatisticWidget;

  beforeEach(() => {
    props = {
      loading: false,
      title: 'Title',
      content: 284,
      inlineContent: false,
    };
  });

  it('should render', () => {
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with secondary values', () => {
    props.subtitle = 'Subtitle';
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with inline content', () => {
    props.inlineContent = true;
    props.renderActivity = <div>renderActivity</div>;
    props.renderAuxContent = <div>renderAuxContent</div>;
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with loading status', () => {
    props.loading = true;
    props.inlineContent = false;
    props.renderAuxContent = <div>renderAuxContent</div>;
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with loading status with no auxContent', () => {
    props.loading = true;
    props.inlineContent = false;
    props.renderActivity = <div>renderActivity</div>;
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should render with loading status and inline content', () => {
    props.loading = true;
    props.inlineContent = true;
    props.renderActivity = <div>renderActivity</div>;
    props.renderAuxContent = <div>renderAuxContent</div>;
    wrapper = render(<StatisticWidget {...props} />);
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
