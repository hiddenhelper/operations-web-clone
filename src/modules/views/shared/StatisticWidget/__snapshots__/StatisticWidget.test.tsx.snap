// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticWidget should render 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-13 makeStyles-noDetail-40 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        class="makeStyles-widgetContainer-14"
      >
        <p
          class="MuiTypography-root makeStyles-widgetTitle-15 makeStyles-widgetTitleBottomMargin-19 MuiTypography-body1"
        >
          Title
        </p>
        <div
          class="makeStyles-widgetContentWrapper-20 makeStyles-widgetContentWrapperBottomMargin-21"
        >
          <span
            class="makeStyles-widgetContentResponsive-26 makeStyles-widgetContent-25"
          >
            284
          </span>
          <div
            class="makeStyles-chipStatusLeftMargin-35"
          />
        </div>
        <div
          class=""
        />
      </div>
      <div
        class="makeStyles-widgetContentWrapperNoTopMargin-22"
      />
    </div>
  </div>
</body>
`;

exports[`StatisticWidget should render with inline content 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-109 makeStyles-noDetail-136 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        class="makeStyles-widgetContainer-110"
      >
        <div>
          <div
            class="makeStyles-widgetInlineStatusWrapper-125"
          >
            <div
              class="makeStyles-widgetContentWrapper-116 "
            >
              <span
                class="makeStyles-widgetContent-121 makeStyles-widgetInlineMargin-123"
              >
                284
              </span>
              <p
                class="MuiTypography-root  makeStyles-widgetTitle-111 makeStyles-widgetTitleInlinePosition-126 MuiTypography-body1"
              >
                Title
              </p>
              <div
                class="makeStyles-chipStatusLeftMargin-131"
              />
            </div>
            <div
              class=""
            >
              <div>
                renderActivity
              </div>
            </div>
          </div>
          <p
            class="MuiTypography-root makeStyles-widgetSubtitle-112 makeStyles-widgetSubtitleInlinePosition-114 MuiTypography-body1"
          />
        </div>
        <div
          class="makeStyles-widgetAuxContentPadding-120"
        >
          <div>
            renderAuxContent
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`StatisticWidget should render with loading status 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-157 makeStyles-noDetail-184 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div>
        <p
          class="MuiTypography-root makeStyles-widgetTitle-159 makeStyles-widgetTitleBottomMargin-163 MuiTypography-body1"
        >
          Title
        </p>
        <div
          class="makeStyles-columnMainContent-147"
          data-testid="statistics-widget-skeleton"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text makeStyles-columnContent-148 MuiSkeleton-wave"
            style="width: 270px; height: 63px;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rect makeStyles-columnAuxContent-149 MuiSkeleton-wave"
            style="width: 100%; height: 168px;"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`StatisticWidget should render with loading status and inline content 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-253 makeStyles-noDetail-280 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div>
        <div
          class="makeStyles-wrapper-241"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text makeStyles-inlineMain-248 MuiSkeleton-wave"
            style="width: 270px; height: 58px;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-text makeStyles-inlineActivity-249 MuiSkeleton-wave"
            style="width: 270px; height: 25px;"
          />
        </div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rect makeStyles-inlineAuxContent-250 MuiSkeleton-wave"
          style="width: 100%; height: 216px;"
        />
        <span
          class="MuiSkeleton-root MuiSkeleton-text makeStyles-inlineCta-251 MuiSkeleton-wave"
          style="width: 120px; height: 23px;"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`StatisticWidget should render with loading status with no auxContent 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-205 makeStyles-noDetail-232 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div>
        <p
          class="MuiTypography-root makeStyles-widgetTitle-207 makeStyles-widgetTitleBottomMargin-211 MuiTypography-body1"
        >
          Title
        </p>
        <div
          class="makeStyles-columnMainContent-195"
          data-testid="statistics-widget-skeleton"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text makeStyles-columnContent-196 MuiSkeleton-wave"
            style="width: 270px; height: 63px;"
          />
        </div>
        <span
          class="MuiSkeleton-root MuiSkeleton-text makeStyles-columnActivity-198 MuiSkeleton-wave"
          style="width: 200px; height: 23px;"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`StatisticWidget should render with secondary values 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root makeStyles-widget-61 makeStyles-noDetail-88 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        class="makeStyles-widgetContainer-62"
      >
        <p
          class="MuiTypography-root makeStyles-widgetTitle-63 makeStyles-widgetTitleBottomMargin-67 MuiTypography-body1"
        >
          Title
        </p>
        <p
          class="MuiTypography-root makeStyles-widgetSubtitle-64 makeStyles-widgetSubtitlePosition-65 MuiTypography-body1"
        >
          Subtitle
        </p>
        <div
          class="makeStyles-widgetContentWrapper-68 makeStyles-widgetTitleMarginBottom-87"
        >
          <span
            class="makeStyles-widgetContentResponsive-74 makeStyles-widgetContent-73"
          >
            284
          </span>
          <div
            class="makeStyles-chipStatusLeftMargin-83"
          />
        </div>
        <div
          class=""
        />
      </div>
      <div
        class="makeStyles-widgetContentWrapperNoTopMargin-70"
      />
    </div>
  </div>
</body>
`;
